scalafmt-core/shared/src/main/scala/org/scalafmt/util/MarkdownFile.scala

```diff
@@ -0,0 +1,133 @@
+package org.scalafmt.util
+
+import scala.meta.inputs.Position
+import scala.meta.inputs.Input
+import scala.collection.mutable
+
+final case class MarkdownFile(input: Input, parts: List[MarkdownPart]) {
+  private val appends = mutable.ListBuffer.empty[String]
+  def appendText(text: String): Unit = {
+    appends += text
+  }
+  def renderToString: String = {
+    val out = new StringBuilder()
+    parts.foreach(_.renderToString(out))
+    appends.foreach(a => out.append(a).append("\n"))
+    out.toString()
+  }
+}
+object MarkdownFile {
+  sealed abstract class State
+  object State {
+    case class CodeFence(start: Int, backticks: String, info: String)
+        extends State
+    case object Text extends State
+  }
+  class Parser(input: Input) {
+    private val text = input.text
+    private def newPos(start: Int, end: Int): Position = {
+      Position.Range(input, start, end)
+    }
+    private def newText(start: Int, end: Int): Text = {
+      val adaptedEnd = math.max(start, end)
+      val part = Text(text.substring(start, adaptedEnd))
+      part.pos = newPos(start, adaptedEnd)
+      part
+    }
+    private def newCodeFence(
+        state: State.CodeFence,
+        backtickStart: Int,
+        backtickEnd: Int
+    ): CodeFence = {
+      val open = newText(state.start, state.start + state.backticks.length())
+      val info = newText(open.pos.end, open.pos.end + state.info.length())
+      val adaptedBacktickStart = math.max(0, backtickStart - 1)
+      val body = newText(info.pos.end, adaptedBacktickStart)
+      val close = newText(adaptedBacktickStart, backtickEnd)
+      val part = CodeFence(open, info, body, close)
+      part.pos = newPos(state.start, backtickEnd)
+      part
+    }
+    def acceptParts(): List[MarkdownPart] = {
+      var state: State = State.Text
+      val parts = mutable.ListBuffer.empty[MarkdownPart]
+      var curr = 0
+      text.linesWithSeparators.foreach { line =>
+        val end = curr + line.length()
+        state match {
+          case State.Text =>
+            if (line.startsWith("```")) {
```

Comment:
what if the fence is not at the beginning of the line?

Reply:
I struggled with this for a few hours and was unable to come up with the correct combination of trimming and character counting to get it to format indented fences.

So at this point, I just want to seek clarification-Is this a requirement? 

I just tried indenting my `scala mdoc` fences in my book project, and that breaks mdoc already.  Or more accurately, mdoc ignores indented fences. Since that's the case, I'm wondering if trying to format indented `scala mdoc` fences is a valid goal.